In [1]:
import pandas as pd
import matplotlib as plt
import numpy as np
pd.set_option('display.max_rows', 50)
pd.set_option('display.max_columns', 500)

## Cleaning the data

* Read the data.
* Select the required features and discard the irrelevant features.
* Clean the relevant features.

In [2]:
df = pd.read_csv('data.csv', index_col=0)
df

ID                Name  Age  \
0      158023            L. Messi   31   
1       20801   Cristiano Ronaldo   33   
2      190871           Neymar Jr   26   
3      193080              De Gea   27   
4      192985        K. De Bruyne   27   
...       ...                 ...  ...   
18202  238813        J. Lundstram   19   
18203  243165  N. Christoffersson   19   
18204  241638           B. Worman   16   
18205  246268      D. Walker-Rice   17   
18206  246269           G. Nugent   16   

                                                Photo Nationality  \
0      https://cdn.sofifa.org/players/4/19/158023.png   Argentina   
1       https://cdn.sofifa.org/players/4/19/20801.png    Portugal   
2      https://cdn.sofifa.org/players/4/19/190871.png      Brazil   
3      https://cdn.sofifa.org/players/4/19/193080.png       Spain   
4      https://cdn.sofifa.org/players/4/19/192985.png     Belgium   
...                                               ...         ...   
18202  https://cdn.sofifa.org/players/4/19/238813.png     England   
18203  https://cdn.sofifa.org/players/4/19/243165.png      Sweden   
18204  https://cdn.sofifa.org/players/4/19/241638.png     England   
18205  https://cdn.sofifa.org/players/4/19/246268.png     England   
18206  https://cdn.sofifa.org/players/4/19/246269.png     England   

                                      Flag  Overall  Potential  \
0      https://cdn.sofifa.org/flags/52.png       94         94   
1      https://cdn.sofifa.org/flags/38.png       94         94   
2      https://cdn.sofifa.org/flags/54.png       92         93   
3      https://cdn.sofifa.org/flags/45.png       91         93   
4       https://cdn.sofifa.org/flags/7.png       91         92   
...                                    ...      ...        ...   
18202  https://cdn.sofifa.org/flags/14.png       47         65   
18203  https://cdn.sofifa.org/flags/46.png       47         63   
18204  https://cdn.sofifa.org/flags/14.png       47         67   
18205  https://cdn.sofifa.org/flags/14.png       47         66   
18206  https://cdn.sofifa.org/flags/14.png       46         66   

                      Club                                       Club Logo  \
0             FC Barcelona    https://cdn.sofifa.org/teams/2/light/241.png   
1                 Juventus     https://cdn.sofifa.org/teams/2/light/45.png   
2      Paris Saint-Germain     https://cdn.sofifa.org/teams/2/light/73.png   
3        Manchester United     https://cdn.sofifa.org/teams/2/light/11.png   
4          Manchester City     https://cdn.sofifa.org/teams/2/light/10.png   
...                    ...                                             ...   
18202      Crewe Alexandra    https://cdn.sofifa.org/teams/2/light/121.png   
18203       Trelleborgs FF    https://cdn.sofifa.org/teams/2/light/703.png   
18204     Cambridge United   https://cdn.sofifa.org/teams/2/light/1944.png   
18205      Tranmere Rovers  https://cdn.sofifa.org/teams/2/light/15048.png   
18206      Tranmere Rovers  https://cdn.sofifa.org/teams/2/light/15048.png   

         Value   Wage  Special Preferred Foot  International Reputation  \
0      €110.5M  €565K     2202           Left                       5.0   
1         €77M  €405K     2228          Right                       5.0   
2      €118.5M  €290K     2143          Right                       5.0   
3         €72M  €260K     1471          Right                       4.0   
4        €102M  €355K     2281          Right                       4.0   
...        ...    ...      ...            ...                       ...   
18202     €60K    €1K     1307          Right                       1.0   
18203     €60K    €1K     1098          Right                       1.0   
18204     €60K    €1K     1189          Right                       1.0   
18205     €60K    €1K     1228          Right                       1.0   
18206     €60K    €1K     1321          Right                       1.0   

       Weak Foot  Skill Moves       Work Rate 

#### **Selecting relevant columns**

We discard irrelevant columns such as "Flag", "Club Logo", etc. There are a lot of columns related to the position of a player on the field (Ex: LS, ST, RS, CF, etc.). These values represent the rating of a player at each position. For example, Messi has a high rating in forward positions and a lower rating when playing in a defensive position. These columns are mostly there to help place the players in the correct position in the field (Ex: you wouldn't want messi in a defensive position). Since our task deals with predicting the rating of a player, these columns aren't useful to our task and we drop them.

We believe that any player, regardless of nationality can reach any skill level. Hence, although we retain certain columns such as "Nationality" "Club", we don't use these to predict the rating as this may result in a biased model.

The following columns rate various skills of a player. Every column contains a value between 0 to 100 with 0 being the lowest rating and 100 being the highest rating.

```
skills = [ 'Crossing', 'Finishing', 'HeadingAccuracy', 
       'ShortPassing', 'Volleys', 'Dribbling',
       'Curve', 'FKAccuracy', 'LongPassing', 'BallControl', 'Acceleration',
       'SprintSpeed', 'Agility', 'Reactions', 'Balance', 'ShotPower',
       'Jumping', 'Stamina', 'Strength', 'LongShots', 'Aggression',
       'Interceptions', 'Positioning', 'Vision', 'Penalties', 'Composure',
       'Marking', 'StandingTackle', 'SlidingTackle', 'GKDiving', 'GKHandling',
       'GKKicking', 'GKPositioning', 'GKReflexes']
```

The following attributes are also used to predict the rating. 'Position' is the players position on the field.

```
xs = [ 'Position', 'Age', 'Height', 'Weight' ]
```

The following column provides the rating of the player. This is what we're trying to predict

```
y = 'Overall'
```

The below fields are kept but aren't used in the prediction. They may be used to interpret the results.

```
extras = ['ID', 'Name',  'Nationality', 'Club']
```

In [3]:
extras = ['ID', 'Name',  'Nationality', 'Club']

y = 'Overall'

xs = [ 'Position', 'Age', 'Height', 'Weight' ]

skills = [ 'Crossing', 'Finishing', 'HeadingAccuracy', 
       'ShortPassing', 'Volleys', 'Dribbling',
       'Curve', 'FKAccuracy', 'LongPassing', 'BallControl', 'Acceleration',
       'SprintSpeed', 'Agility', 'Reactions', 'Balance', 'ShotPower',
       'Jumping', 'Stamina', 'Strength', 'LongShots', 'Aggression',
       'Interceptions', 'Positioning', 'Vision', 'Penalties', 'Composure',
       'Marking', 'StandingTackle', 'SlidingTackle', 'GKDiving', 'GKHandling',
       'GKKicking', 'GKPositioning', 'GKReflexes']

all_cols = extras + xs + [y] + skills
len(all_cols)

43

In [4]:
# we keep only the selected columns and drop everything else
df = df[all_cols]
df

ID                Name Nationality                 Club Position  \
0      158023            L. Messi   Argentina         FC Barcelona       RF   
1       20801   Cristiano Ronaldo    Portugal             Juventus       ST   
2      190871           Neymar Jr      Brazil  Paris Saint-Germain       LW   
3      193080              De Gea       Spain    Manchester United       GK   
4      192985        K. De Bruyne     Belgium      Manchester City      RCM   
...       ...                 ...         ...                  ...      ...   
18202  238813        J. Lundstram     England      Crewe Alexandra       CM   
18203  243165  N. Christoffersson      Sweden       Trelleborgs FF       ST   
18204  241638           B. Worman     England     Cambridge United       ST   
18205  246268      D. Walker-Rice     England      Tranmere Rovers       RW   
18206  246269           G. Nugent     England      Tranmere Rovers       CM   

       Age Height  Weight  Overall  Crossing  Finishing  HeadingAccuracy  \
0       31    5'7  159lbs       94      84.0       95.0             70.0   
1       33    6'2  183lbs       94      84.0       94.0             89.0   
2       26    5'9  150lbs       92      79.0       87.0             62.0   
3       27    6'4  168lbs       91      17.0       13.0             21.0   
4       27   5'11  154lbs       91      93.0       82.0             55.0   
...    ...    ...     ...      ...       ...        ...              ...   
18202   19    5'9  134lbs       47      34.0       38.0             40.0   
18203   19    6'3  170lbs       47      23.0       52.0             52.0   
18204   16    5'8  148lbs       47      25.0       40.0             46.0   
18205   17   5'10  154lbs       47      44.0       50.0             39.0   
18206   16   5'10  176lbs       46      41.0       34.0             46.0   

       ShortPassing  Volleys  Dribbling  Curve  FKAccuracy  LongPassing  \
0              90.0     86.0       97.0   93.0        94.0         87.0   
1              81.0     87.0       88.0   81.0        76.0         77.0   
2              84.0     84.0       96.0   88.0        87.0         78.0   
3              50.0     13.0       18.0   21.0        19.0         51.0   
4              92.0     82.0       86.0   85.0        83.0         91.0   
...             ...      ...        ...    ...         ...          ...   
18202          49.0     25.0       42.0   30.0        34.0         45.0   
18203          43.0     36.0       39.0   32.0        20.0         25.0   
18204          38.0     38.0       45.0   38.0        27.0         28.0   
18205          42.0     40.0       51.0   34.0        32.0         32.0   
18206          48.0     30.0       43.0   40.0        34.0         44.0   

       BallControl  Acceleration  SprintSpeed  Agility  Reactions  Balance  \
0             96.0          91.0         86.0     91.0       95.0     95.0   
1             94.0          89.0         91.0     87.0       96.0     70.0   
2             95.0          94.0         90.0     96.0       94.0     84.0   
3             42.0          57.0         58.0     60.0       90.0     43.0   
4             91.0          78.0         76.0     79.0       91.0     77.0   
...            ...           ...          ...      ...        ...      ...   
18202         43.0          54.0         57.0     60.0       49.0     76.0   
18203         40.0          41.0         39.0     38.0       40.0     52.0   
18204         44.0          70.0         69.0     50.0       47.0     58.0   
18205         52.0          61.0         60.0     52.0       21.0     71.0   
18206         51.0          57.0         55.0     55.0       51.0     63.0   

       ShotPower  Jumping  Stamina  Strength  LongShots  Aggression  \
0           85.0     68.0     72.0      59.0       94.0        48.0   
1           95.0     95.0     88.0      79.0       93.0        63.0   
2           80.0     61.0     81.0      49.0       82.0        56.0   
3           31.0     67.0  

#### **Handling missing data**

We check each column to see if we have missing data

In [5]:
ser = df.isna().sum()
ser = ser.sort_values()
for i in range(len(ser)):
    print(ser.index[i], '\t\t', ser.iloc[i])

ID 		 0
Name 		 0
Nationality 		 0
Age 		 0
Overall 		 0
Jumping 		 48
Stamina 		 48
Strength 		 48
LongShots 		 48
Aggression 		 48
Interceptions 		 48
Positioning 		 48
Vision 		 48
Composure 		 48
ShotPower 		 48
Marking 		 48
StandingTackle 		 48
SlidingTackle 		 48
GKDiving 		 48
GKHandling 		 48
GKKicking 		 48
Penalties 		 48
Balance 		 48
Agility 		 48
GKPositioning 		 48
Height 		 48
Weight 		 48
Crossing 		 48
Finishing 		 48
HeadingAccuracy 		 48
Reactions 		 48
Volleys 		 48
ShortPassing 		 48
Curve 		 48
FKAccuracy 		 48
LongPassing 		 48
BallControl 		 48
Acceleration 		 48
SprintSpeed 		 48
Dribbling 		 48
GKReflexes 		 48
Position 		 60
Club 		 241


We try dropping all the rows with missing data.

In [6]:
df = df.dropna()
df.shape

(17918, 43)

In [7]:
18207 - 17918

289

We end up dropping 289 rows. This seems reasonable given that we still have plenty of data to work with.

#### **Feature engineering and cleaning**

The following tasks need to be completed - 

1. Position - this can be reduced to just "Forward", "Middle", "Defense", and "Goalkeeper" and encode it numerically.
2. Height - needs to be converted from feet and inches to just feet.
3. Weight - the "lbs" at the end needs to be removed and the column needs to be converted to a floating point data type.

**Position**

The below picture explains the positions of players in a football field (Source - https://www.stack.com/a/the-soccer-positions-explained/)

![](./soccer-positions.webp)

In [8]:
sorted(df.Position.unique())

['CAM',
 'CB',
 'CDM',
 'CF',
 'CM',
 'GK',
 'LAM',
 'LB',
 'LCB',
 'LCM',
 'LDM',
 'LF',
 'LM',
 'LS',
 'LW',
 'LWB',
 'RAM',
 'RB',
 'RCB',
 'RCM',
 'RDM',
 'RF',
 'RM',
 'RS',
 'RW',
 'RWB',
 'ST']

* Goalkeeper - "GK"
* Midfielder - anything ending with "M"
* Defense - anything ending with "B" ("SW" doesn't exist in our dataset)
* Forward - anything ending with "F", "S", "T", or "W"

In [9]:
def reduce_positions(pos):
    if pos == "GK":
        return "Goalkeeper"
    
    pos = pos.lower().strip()
    
    if pos.endswith("m"):
        return "Midfield"
    
    if pos.endswith("b"):
        return "Defense"
    
    if pos[-1] in ['f', 's', 't', 'w']:
        return "Forward"
    
    return None

In [10]:
df['precise_position'] = df.Position
df.loc[:, 'Position'] = df.Position.apply(reduce_positions)
df.Position.isna().sum()

C:\Users\geeth\AppData\Local\Temp/ipykernel_135604/3056951590.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['precise_position'] = df.Position
C:\Users\geeth\anaconda3\lib\site-packages\pandas\core\indexing.py:1773: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self._setitem_single_column(ilocs[0], value, pi)


0

In [11]:
df.Position.value_counts()

Midfield      6764
Defense       5781
Forward       3381
Goalkeeper    1992
Name: Position, dtype: int64

We use one hot encoding to encode the positions since none of the values have any ordering.

In [12]:
pos_df = pd.get_dummies(df.Position)
df = pd.concat((df.drop(columns=['Position']), pos_df), axis=1)
df

ID                Name Nationality                 Club  Age  \
0      158023            L. Messi   Argentina         FC Barcelona   31   
1       20801   Cristiano Ronaldo    Portugal             Juventus   33   
2      190871           Neymar Jr      Brazil  Paris Saint-Germain   26   
3      193080              De Gea       Spain    Manchester United   27   
4      192985        K. De Bruyne     Belgium      Manchester City   27   
...       ...                 ...         ...                  ...  ...   
18202  238813        J. Lundstram     England      Crewe Alexandra   19   
18203  243165  N. Christoffersson      Sweden       Trelleborgs FF   19   
18204  241638           B. Worman     England     Cambridge United   16   
18205  246268      D. Walker-Rice     England      Tranmere Rovers   17   
18206  246269           G. Nugent     England      Tranmere Rovers   16   

      Height  Weight  Overall  Crossing  Finishing  HeadingAccuracy  \
0        5'7  159lbs       94      84.0       95.0             70.0   
1        6'2  183lbs       94      84.0       94.0             89.0   
2        5'9  150lbs       92      79.0       87.0             62.0   
3        6'4  168lbs       91      17.0       13.0             21.0   
4       5'11  154lbs       91      93.0       82.0             55.0   
...      ...     ...      ...       ...        ...              ...   
18202    5'9  134lbs       47      34.0       38.0             40.0   
18203    6'3  170lbs       47      23.0       52.0             52.0   
18204    5'8  148lbs       47      25.0       40.0             46.0   
18205   5'10  154lbs       47      44.0       50.0             39.0   
18206   5'10  176lbs       46      41.0       34.0             46.0   

       ShortPassing  Volleys  Dribbling  Curve  FKAccuracy  LongPassing  \
0              90.0     86.0       97.0   93.0        94.0         87.0   
1              81.0     87.0       88.0   81.0        76.0         77.0   
2              84.0     84.0       96.0   88.0        87.0         78.0   
3              50.0     13.0       18.0   21.0        19.0         51.0   
4              92.0     82.0       86.0   85.0        83.0         91.0   
...             ...      ...        ...    ...         ...          ...   
18202          49.0     25.0       42.0   30.0        34.0         45.0   
18203          43.0     36.0       39.0   32.0        20.0         25.0   
18204          38.0     38.0       45.0   38.0        27.0         28.0   
18205          42.0     40.0       51.0   34.0        32.0         32.0   
18206          48.0     30.0       43.0   40.0        34.0         44.0   

       BallControl  Acceleration  SprintSpeed  Agility  Reactions  Balance  \
0             96.0          91.0         86.0     91.0       95.0     95.0   
1             94.0          89.0         91.0     87.0       96.0     70.0   
2             95.0          94.0         90.0     96.0       94.0     84.0   
3             42.0          57.0         58.0     60.0       90.0     43.0   
4             91.0          78.0         76.0     79.0       91.0     77.0   
...            ...           ...          ...      ...        ...      ...   
18202         43.0          54.0         57.0     60.0       49.0     76.0   
18203         40.0          41.0         39.0     38.0       40.0     52.0   
18204         44.0          70.0         69.0     50.0       47.0     58.0   
18205         52.0          61.0         60.0     52.0       21.0     71.0   
18206         51.0          57.0         55.0     55.0       51.0     63.0   

       ShotPower  Jumping  Stamina  Strength  LongShots  Aggression  \
0           85.0     68.0     72.0      59.0       94.0        48.0   
1           95.0     95.0     88.0      79.0       93.0        63.0   
2           80.0     61.0     81.0      49.0       82.0        56.0   
3           31.0     67.0     43.0      64.0       12.0        38.0   
4           91.0     63.0     90.0      75.0       91.0        7

**Height** 

We convert feet and inches to just feet. There are 12 inches in a feet. We divide the inches by 12 and add it to the feet.

In [13]:
# verifying that all heights have the same format
df.Height.value_counts()

6'0     2837
5'10    2452
5'9     2203
5'11    2132
6'2     1990
6'1     1886
6'3      975
5'8      934
5'7      892
6'4      741
5'6      311
6'5      241
5'5      144
6'6       91
5'4       30
6'7       21
5'3       18
6'8       10
5'2        5
5'1        3
6'9        2
Name: Height, dtype: int64

In [14]:
def convert_to_feet(val):
    val = val.split("'")
    return float(val[0]) + (float(val[1]) / 12.0)

In [15]:
df.loc[:, 'Height'] = df.Height.apply(convert_to_feet)
df.Height

0        5.583333
1        6.166667
2        5.750000
3        6.333333
4        5.916667
           ...   
18202    5.750000
18203    6.250000
18204    5.666667
18205    5.833333
18206    5.833333
Name: Height, Length: 17918, dtype: float64

**Weight**

Removing "lbs" from the Weight column

In [16]:
df.Weight.str.endswith('lbs').sum(), df.shape[0]

(17918, 17918)

All values end with "lbs". Hence, we can just remove the last 3 characters.

In [17]:
df.loc[:, 'Weight'] = df.Weight.str[:-3].astype('float32')
df.Weight

0        159.0
1        183.0
2        150.0
3        168.0
4        154.0
         ...  
18202    134.0
18203    170.0
18204    148.0
18205    154.0
18206    176.0
Name: Weight, Length: 17918, dtype: float32

**Save preprocessed dataset** 

In [18]:
df.to_csv('preproc.csv', index=False)